### ISSUE TITLE
Fine Tune other models such as ResNet50

### ISSUE THREAD

**Issue by abdualhag** [deeplizard/Keras_Jupyter_Notebooks#3]
Hi,
I just discovered your channel and needless to say, great tutorials. I worked out your fine tuning using VGG16 but once I started testing on other models, I ran to various issues. 

Below what works:
```
import keras
from keras import backend as K
from keras.models import Sequential, Model
from keras.layers import Activation, Dense
from keras.optimizers import Adam
from keras.metrics import categorical_crossentropy
from keras.preprocessing.image import ImageDataGenerator
from keras.preprocessing import image
from keras.applications.vgg16 import VGG16

train_path = 'cats-and-dogs/train'
valid_path = 'cats-and-dogs/valid'
test_path = 'cats-and-dogs/test'


# dimensions of our images.
img_width, img_height = 224, 224

if K.image_data_format() == 'channels_first':
    input_shape = (3, img_width, img_height)
else:
    input_shape = (img_width, img_height, 3)


train_batches = ImageDataGenerator().flow_from_directory(train_path, target_size=(img_width,img_height), classes=['dog', 'cat'], batch_size=10)
valid_batches = ImageDataGenerator().flow_from_directory(valid_path, target_size=(img_width,img_height), classes=['dog', 'cat'], batch_size=4)


vgg16_model = VGG16()

model = Sequential()
for layer in vgg16_model.layers:
    model.add(layer)

model.layers.pop()
for layer in model.layers:
    layer.trainable = False
model.add(Dense(2, activation='softmax'))

model.compile(Adam(lr=.0001), loss='categorical_crossentropy', metrics=['accuracy'])

model.fit_generator(train_batches, steps_per_epoch=4, 
                    validation_data=valid_batches, validation_steps=4, epochs=5, verbose=2)

```

According to Keras documentation, the following is how you can fine tune InceptionV3.
```


from keras.applications.inception_v3 import InceptionV3
from keras.preprocessing import image
from keras.models import Model
from keras.layers import Dense, GlobalAveragePooling2D
from keras.preprocessing.image import ImageDataGenerator
from keras import backend as K

train_path = 'cats-and-dogs/train'
valid_path = 'cats-and-dogs/valid'
test_path = 'cats-and-dogs/test'

img_width, img_height = 299, 299

if K.image_data_format() == 'channels_first':
    input_shape = (3, img_width, img_height)
else:
    input_shape = (img_width, img_height, 3)


train_batches = ImageDataGenerator().flow_from_directory(train_path, target_size=(img_width,img_height), classes=['dog', 'cat'], batch_size=10)
valid_batches = ImageDataGenerator().flow_from_directory(valid_path, target_size=(img_width,img_height), classes=['dog', 'cat'], batch_size=4)

base_model = InceptionV3(weights='imagenet', include_top=False)


# add a global spatial average pooling layer
x = base_model.output
x = GlobalAveragePooling2D()(x)
# let's add a fully-connected layer
x = Dense(1024, activation='relu')(x)
# and a logistic layer -- we have 2 classes
predictions = Dense(2, activation='softmax')(x)

# this is the model we will train
model = Model(inputs=base_model.input, outputs=predictions)


for layer in base_model.layers:
    layer.trainable = False


# compile the model (should be done *after* setting layers to non-trainable)
model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])


# train the model on the new data for a few epochs
#model.fit_generator(...)

model.fit_generator(
    train_generator,
    steps_per_epoch=4,
    epochs=5,
    validation_data=validation_generator,
    validation_steps=4)

```
The code above compile but the accuracy never go above 0.5. When I tried to create sequential model and add the InceptionV3 layers to it as you did with VGG16, it simply would not compile.

Giving your expertise, I thought you might be able to help. 
Looking forward to hear what you got to say. Thanks.

**Comment by deeplizard** [comment]
Hey abdualhag, I'm glad you're enjoying the tutorials. Thanks for letting me know!

In regards to adding layers from InceptionV3 to a Sequential model-- Another viewer previously mentioned the following after she also tried to do the same: "_Since Xception and InceptionV3 models have 'branch nodes', we can't add each individual layer into a Sequential container. This is why I was getting the error message. Instead, the recommended approach was to instantiate a InceptionV3 or Xception model with the parameter include_top = False which excludes the final output layer﻿._"

In regards to why the accuracy doesn't rise above 0.5 whenever using the code provided by Keras for InceptionV3-- According to the Inceptionv3 code from Keras ([here](https://github.com/keras-team/keras/blob/master/keras/applications/inception_v3.py)), there is a preprocessing function that you have to run your data through. It starts on line 398. It doesn't appear that you are preprocessing your image data accordingly for this model.

If you'd like more info, here is a stackoverflow Q&A about the preprocessing required for Inception V3.
https://stackoverflow.com/questions/44341258/preprocessing-function-of-inception-v3-in-keras

**Comment by abdualhag** [comment]
Thanks for getting back to me. As for preprocessing the data, I am quiet confused as to where I call the preprocessor function and on what? Here is my naive try which does not work:

```
def preprocess_input(x):
    x /= 255.
    x -= 0.5
    x *= 2.
    return x 

train_batches = preprocess_input(ImageDataGenerator().flow_from_directory(train_path, target_size=(img_width,img_height), classes=['dog', 'cat'], batch_size=10))
valid_batches = preprocess_input(ImageDataGenerator().flow_from_directory(valid_path, target_size=(img_width,img_height), classes=['dog', 'cat'], batch_size=4))
```

Another thing I tried to do is the following which compiles but does not raise the accuracy above 0.5, I do not think I am changing the right variables:

```
# prepare data augmentation configuration
train_datagen = ImageDataGenerator(
    rescale=1. / 255,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True)

test_datagen = ImageDataGenerator(rescale=1. / 255)

batches = train_datagen.flow_from_directory(train_path, target_size=(img_width,img_height), classes=['dog', 'cat'], batch_size=10)
valid_batches = test_datagen.flow_from_directory(valid_path, target_size=(img_width,img_height), classes=['dog', 'cat'], batch_size=4)
```

I would love to hear your thoughts. Thanks

**Comment by deeplizard** [comment]
I've not fine-tuned this model myself, but the comments on the code for this preprocessing function state, ```
# Preprocesses a numpy array encoding a batch of images.
    # Arguments
        x: a 4D numpy array consists of RGB values within [0, 255].
```

So, it appears you should b passing your image batches to it before you pass those batches to the fit_generator() function.

**Comment by abdualhag** [comment]
Sound good. Thanks
